### PR TITLE
Resolve circular dependency

### LINKED ISSUE #545: Circular dependency issue
There's a circular dependency whenever we import `util.py` in other modules like `store.py`, etc.

```
ImportError while loading conftest '/jupysql/src/tests/conftest.py'.
src/tests/conftest.py:8: in <module>
    from sql.magic import SqlMagic, RenderMagic
../../miniforge3/envs/jupysql/lib/python3.10/site-packages/sql/__init__.py:1: in <module>
    from .magic import RenderMagic, SqlMagic, load_ipython_extension
../../miniforge3/envs/jupysql/lib/python3.10/site-packages/sql/magic.py:22: in <module>
    import sql.connection
../../miniforge3/envs/jupysql/lib/python3.10/site-packages/sql/connection.py:11: in <module>
    from sql.store import store
../../miniforge3/envs/jupysql/lib/python3.10/site-packages/sql/store.py:10: in <module>
    from sql import util
../../miniforge3/envs/jupysql/lib/python3.10/site-packages/sql/util.py:3: in <module>
    from sql import inspect
../../miniforge3/envs/jupysql/lib/python3.10/site-packages/sql/inspect.py:4: in <module>
    from sql.connection import Connection
E   ImportError: cannot import name 'Connection' from partially initialized module 'sql.connection' (most likely due to a circular import) (/Users/neelashasen/miniforge3/envs/jupysql/lib/python3.10/site-packages/sql/connection.py)
```

### LINKED ISSUE #749: unclear logic when handling errors
it is unclear what the logic should be in `error_message.py`.

there is a `_is_syntax_error` function that's supposedly capturing syntax errors, but it also captures things like "catalog error", "not found", and "does not exist", which doesn't seem syntax errors, but rather seem like errors displayed when the user tries to query a non-existent table (which is different than a SQL syntax error).

Since `_is_syntax_error` is used to determine whether to show the "If using snippets, you may pass the --with argument..." error message suggestion, it seems that the purpose of the function is to capture "missing table errors" as opposed to "syntax errors". In such a scenario, we shouldn't be capturing syntax errors.

I modified `test_sql_error_suggests_using_cte` in https://github.com/ploomber/jupysql/pull/746 to reflect this. some tests are marked as xfail because I'm unsure what the purpose of this code is.

### PR DESCRIPTION
**Pull request by neelasha23**
## Describe your changes
1. Moved the `store` utility functions like getting all saved snippets, finding dependents of key, etc to `store_utils.py`, and corresponding tests moved to `test_store.py`.
2. All util functions that access current connection moved from `util.py` to `inspect.py`, and corresponding tests moved to `test_inspect.py`
3. Renamed `error_messages.py` to `error_handler.py` and structured the code by adding more internal functions.
4. Moved logic of extracting table names from query to `util.py`.
5. Added functions to check if error is sqlalchemy error or not, and substring check in `util.py`. this is needed for the detailed error message logic.
6. Handling of DuckDB Parser error and pyodbc.ProgrammingError added. so some of the integration tests failing previously have now been fixed.

## Issue number

Closes #749  #545 

## Checklist before requesting a review

- [ ] Performed a self-review of my code
- [ ] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)
- [ ] Added [tests](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#testing) (when necessary).
- [ ] Added [docstring](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#documenting-changes-and-new-features) documentation and update the [changelog](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#changelog) (when needed)

----
:books: Documentation preview :books:: https://jupysql--787.org.readthedocs.build/en/787/

### CODE REVIEW

**Comment by neelasha23** [comment]
Please review @edublancas

**Comment by edublancas** [comment]
@neelasha23 I was about to review this but now I see that the tests are failing. what's the status? I guess there are merge conflicts due to recent changes in master?

**Comment by neelasha23** [comment]
> @neelasha23 I was about to review this but now I see that the tests are failing. what's the status? I guess there are merge conflicts due to recent changes in master?

Yes the snippets related tests started failing after I rebased. Fixed now. Please review.
@edublancas

**Comment by neelasha23** [comment]
Please review @edublancas

**Review by edublancas — CHANGES_REQUESTED** [review]
I just realized that I had reviewed this before but never submitted my observations  (they still appeared as "pending" but I think that means you'll never saw them), so I'm submitting them now.

there's a lot of changes in this PR so please write a brief description so I know what to expect. e.g., "moved functions from a.py, b.py, moved functions from c.py into a separate module"

i guess this is mostly moving stuff around and not adding any new stuff so please explain

File: src/sql/store_utils.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
+from sql.store import store, get_dependents_for_key
```
[the remark this answers is not in the document]
Comment (neelasha23) [reply src/sql/store_utils.py:1]: yes when we add these functions in `store.py` the tests fail. In `isolate_tests` fixture there's the store is being cleared: `store.store = store.SQLStore()`. This is causing the store to become empty when running each test, although we are creating snippets again in the test.
I'm not sure exactly why this is happening but moving to a separate file solves the issue.

File: src/tests/test_inspect.py [hunk 2]
```diff
@@ -89,6 +96,48 @@ def test_get_column(sample_db, name, first, second, schema):
     assert second in columns._repr_html_()
 
 
+@pytest.mark.parametrize(
```
[the remark this answers is not in the document]
Comment (neelasha23) [reply src/tests/test_inspect.py:99]: yes these are old tests moved from `test_util.py`

File: src/tests/test_magic_cte.py [hunk 3]
```diff
@@ -63,7 +63,11 @@ def test_infer_dependencies(ip, capsys):
 TABLE_NAME_TYPO_ERR_MSG = """
 There is no table with name 'author_subb'.
 Did you mean : 'author_sub'
-If you need help solving this issue, send us a message: https://ploomber.io/community
+
+
+Original error message from DB driver:
+(sqlite3.OperationalError) no such table: author_subb
+[SQL: SELECT last_name FROM author_subb;]
 """
```
[the remark this answers is not in the document]
Comment (neelasha23) [reply src/tests/test_magic_cte.py:71]: previously it was in `magic.py` due to the circular dependency. Now it's move to the error handler file and similar to the syntax error it displays descriptive error message + original db error. original db error was not there previously.

File: CHANGELOG.md [hunk 4]
```diff
@@ -12,6 +12,8 @@
 * [Fix] `ResultSet` footer only displayed when `feedback=2`
 * [Fix] Current connection and switching connections message only displayed when `feedback>=1`
 * [Fix] `--persist/--persist-replace` perform `ROLLBACK` automatically when needed
+* [Fix] Resolve circular dependency issue occurring due to multiple imports in `util.py`
```
[the remark this answers is not in the document]
Comment (neelasha23) [reply CHANGELOG.md]: [fixed](https://github.com/ploomber/jupysql/pull/787/commits/10a987fd245032cae516c5719ec4ebf5c5f77d32)

File: CHANGELOG.md [hunk 5]
```diff
@@ -12,6 +12,8 @@
...
+* [Fix] Handling of DuckDB `Parser Error` and `pyodbc.ProgrammingError`
```
[the remark this answers is not in the document]
Comment (neelasha23) [reply CHANGELOG.md]: [fixed](https://github.com/ploomber/jupysql/pull/787/commits/10a987fd245032cae516c5719ec4ebf5c5f77d32)

File: src/tests/test_inspect.py [hunk 6]
```diff
@@ -11,6 +11,13 @@
 from sql import inspect, connection
 
 
+EXPECTED_SUGGESTIONS_MESSAGE = "Did you mean :"
```
[the remark this answers is not in the document]
Comment (neelasha23) [reply src/tests/test_inspect.py]: [fixed](https://github.com/ploomber/jupysql/pull/787/commits/10a987fd245032cae516c5719ec4ebf5c5f77d32)

File: src/sql/connection/connection.py [hunk 7]
```diff
@@ -858,11 +858,7 @@ def _start_sqlalchemy_connection(cls, engine, connect_str):
             connection = engine.connect()
             return connection
         except OperationalError as e:
-            detailed_msg = detail(e)
-            if detailed_msg is not None:
-                raise exceptions.UsageError(detailed_msg)
-            else:
-                print(e)
+            ErrorHandler(e).handle_exception()
```
[the remark this answers is not in the document]
Comment (neelasha23) [reply src/sql/connection/connection.py]: [fixed](https://github.com/ploomber/jupysql/pull/787/commits/587279506d14c0d97a70f41c0efab07f7029eb7c)

**Comment by neelasha23** [comment]
Addressed your comments and updated the PR description @edublancas

**Review by edublancas — APPROVED** [review]

**Comment by edublancas** [comment]
@neelasha23 good work!

when writing your initial comment, write each issue on a separate line:

closes #X
closes #Y

because if you do

closes #X #Y

github won't close both issues

**Pull request merged**